Add vector-valued variants

tcmoore3 opened the pull request:
## Description

This PR implements a `VectorVariant` that returns array values as a function of time step, in contrast to the `Variant` classes that return scalar values as a function of the time step. `VectorVariant` is templated on the length of the array to take advantage of C++'s strong typing. This PR also implements a `VectorVariantBox` and a handful of subclasses of `VectorVariantBox` that return length-6 arrays for use with functions that take box arguments.

The new `VectorVariant*` C++ classes are implemented in `VectorVariant.[h,cc]`. In Python, I created a new `variant` package. The existing (scalar-valued) `Variant`s are defined in the package's `__init__.py` file, and the box vector variant python classes are in `hoomd/variant/box.py`.

## Motivation and context

This functionality is a first step towards fixing several issues, #1640 and #1641. It will also allow us to replace the arguments `box1`, `box2`, and `variant` with a single `hoomd.variant.box.*` argument to make the use of `update.BoxResize` more intuitive.

## How has this been tested?

Existing unit tests pass, and my ad-hoc tests of the new box vector variant classes are passing. I still need to add proper unit tests for the vector variants.

## Change log

```
- Add `variant.box` classes that describe boxes that change with time step.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

tcmoore3 committed:
Add VectorVariant class

Adds a VectorVariant class.
Adds VectorVariantBox as a subclass of VectorVariant.
Implements VectorVariantBoxConstant and VectorVariantBoxLinear.
Moves variant from a module to a package.

The code compiles and I am able to get the expected boxes as a function
of time for both VectorVariantBoxConstant and VectorVariantBoxLinear
objects.

More to come before I open a PR.

tcmoore3 committed:
Add a linear density ramp box variant

tcmoore3 committed:
First bit of sphinx docs for box variants

tcmoore3 committed:
More sphinx documentation

tcmoore3 committed:
Run pre-commit

tcmoore3 committed:
Fix references in sphinx-doc

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,361 @@
+class PYBIND11_EXPORT VectorVariantBox : public VectorVariant<6>
+    {
+    protected:
+    std::array<Scalar, 6> box_to_array(std::shared_ptr<BoxDim> box)
```
```suggestion
    static std::array<Scalar, 6> box_to_array(std::shared_ptr<BoxDim> box)
```

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,361 @@
+    /// Return the value.
+    virtual array_type operator()(uint64_t timestep)
+        {
+        return m_value;
```
With the `box_to_array` helper above, we should remove `m_value` entirely and replace it with `return box_to_array(m_box)`.

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,361 @@
+    */
+    VectorVariantBoxLinear(std::shared_ptr<BoxDim> box1,
+                           std::shared_ptr<BoxDim> box2,
+                           uint64_t t_start,
```
To provide backwards compatibility with `update.BoxResize`, this should take `std::shared_ptr<Variant>` as an argument instead of `t_start`, `t_stop`.

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,361 @@
+
+    virtual array_type operator()(uint64_t timestep)
+        {
+        if (timestep < m_t_start)
```
To avoid code duplication, use an internal VariantRamp to compute the scale factor. There should be no need for 3 code paths in operator() as the interpolation will naturally range from 0 to 1.

joaander commented on hoomd/__init__.py:
```diff
@@ -61,6 +61,7 @@
 from hoomd import trigger
 from hoomd import variant
 from hoomd.box import Box, box_like
+from hoomd.variant import box
```
```suggestion
```
Keep the package hierarchy. The names should be `hoomd.variant.box.SomeBoxVariant`, not `hoomd.box.SomeBoxVariant`.

joaander commented on hoomd/update/box_resize.py:
```diff
@@ -185,7 +185,7 @@ def get_box(self, timestep):
 
         Returns:
-            Box: The box used at the given timestep.
+            hoomd.Box: The box used at the given timestep.
```
Why all these changes? Sphinx knows we are in the hoomd module, so the `Box` link resolves correctly: https://hoomd-blue.readthedocs.io/en/v4.4.1/module-hoomd-update.html#hoomd.update.BoxResize.get_box

tcmoore3 replied:
I get warnings when I keep it as `Box`:
```
hoomd-blue/hoomd/update/box_resize.py:docstring of hoomd.update.box_resize.BoxResize.get_box:1: WARNING: more than one target found for cross-reference 'Box': hoomd.variant.box.Box, hoomd.Box, hoomd.box.Box
```
and the documentation generated for `hoomd.update.BoxResize.get_box()` links to `hoomd.variant.box.Box` in this case. Is there a way to ensure sphinx finds the right `Box`?

joaander commented on hoomd/variant/__init__.py:
```diff
@@ -7,7 +7,7 @@
+See `Variant` for details on creating user-defined variants or use one of the
 provided subclasses.
 """
 import typing
```
```suggestion
import typing
from . import box
```

joaander commented on sphinx-doc/module-hoomd-variant.rst:
```diff
@@ -39,3 +39,13 @@ hoomd.variant
         :members: min, max, __getstate__, __setstate__
     .. autodata:: variant_like
+
+.. autosummary::
+    :nosignatures:
```
```suggestion
```
No need for an empty auto summary.

joaander reviewed (changes requested):
Thanks! Here are some suggestions to improve this. 

As you already noted, we also need to add unit tests.

tcmoore3 committed:
Apply suggestions from code review

Co-authored-by: Joshua A. Anderson <joaander@umich.edu>

tcmoore3 committed:
Get rid of m_value in constant box variant

tcmoore3 committed:
Replace t_start/t_ramp with a vaiant in VectorVariantBoxLinear

tcmoore3 committed:
Fix imports

tcmoore3 committed:
Attempt to fix imports

melodyyzh commented on hoomd/variant/box.py:
```diff
@@ -0,0 +1,94 @@
+    """
+
+    def __init__(self, initial_box, final_box, variant):
+        box1 = box_preprocessing(box1)
```
Should this be `box1 = box_preprocessing(initial_box)` and `box2 = box_preprocessing(final_box)`? 
This might be the cause for the error in pre-commit.

tcmoore3 replied:
Nice catch!

tcmoore3 committed:
Move scalar variants to new module

Moves the scalar variant classes to a separate module, but imports them
in hoomd/variant/__init__.py so that they still live in the
hoomd.variant namespace. I did this as a workaround for python's import
behavior since I kept getting circular import errors when keeping the
scalar variant classes in the package's __init__.py file.

tcmoore3 committed:
Use internal variant for inverse volume box ramp

tcmoore3 committed:
Call box_preprocessing on correct boxes

tcmoore3 committed:
Fix box references in python docstrings

tcmoore3 committed:
Fix order of imports

tcmoore3 committed:
Import variant.box

tcmoore3 committed:
Update docstring

tcmoore3 committed:
Format

tcmoore3 committed:
Add docstring to variant/scalar.py

tcmoore3 committed:
Rename LinearInverseVolume to InverseVolumeRamp

tcmoore3 committed:
Start implementing tests

tcmoore3 committed:
Add more tests

tcmoore3 committed:
Actually add test file

tcmoore3 committed:
Test custom box variant

tcmoore3 committed:
formatting

tcmoore3 committed:
formatting

tcmoore3 committed:
Fix error in docs and formatting hack

tcmoore3 committed:
Remove pybind11 pickle code

tcmoore3 committed:
Merge remote-tracking branch 'origin/trunk-minor' into feature/vector-variant

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,289 @@
+// Copyright (c) 2009-2023 The Regents of the University of Michigan.
+// Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+// Copyright (c) 2009-2023 The Regents of the University of Michigan.
+// Part of HOOMD-blue, released under the BSD 3-Clause License.
+
```
```suggestion
```
Remove duplicate comment.

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,289 @@
+class PYBIND11_EXPORT VectorVariantBoxInterpolate : public VectorVariantBox
+    {
+    public:
+    /** Construct a VectorVariantBoxInterpolate to interpolate between two boxes linearly in time.
```
```suggestion
    /** Construct a VectorVariantBoxInterpolate to interpolate between two boxes linearly.
```

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,289 @@
+    Scalar m_current_volume;
+
+    /// Variant for computing scale value
+    VariantRamp m_variant; //!< Variant that interpolates between boxes
```
```suggestion
    VariantRamp m_variant;
```
No need to document the member variable twice.

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,289 @@
+    /// Variant that interpolates between boxes.
+    std::shared_ptr<Variant> m_variant;
+    };
+
```
Class level overview documentation for this and the other classes would be useful.

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,289 @@
+class PYBIND11_EXPORT VectorVariantBoxInverseVolumeRamp : public VectorVariantBox
+    {
+    public:
+    VectorVariantBoxInverseVolumeRamp(std::shared_ptr<BoxDim> box1,
+                                      Scalar final_volume,
+                                      uint64_t t_start,
+                                      uint64_t t_ramp)
+        : m_initial_box(box1), m_final_volume(final_volume), m_variant(0, 1, t_start, t_ramp)
```
```suggestion
    VectorVariantBoxInverseVolumeRamp(std::shared_ptr<BoxDim> initial_box,
                                      Scalar final_volume,
                                      uint64_t t_start,
                                      uint64_t t_ramp)
        : m_initial_box(initial_box), m_final_volume(final_volume), m_variant(0, 1, t_start, t_ramp)
```

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,289 @@
+    /// Whether box1 is 2-dimensional or not
+    bool m_is2D;
+
+    /// The current value of the volume
+    Scalar m_current_volume;
```
```suggestion
```
Unused.

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,289 @@
+    Scalar m_final_volume;
+
+    /// Whether box1 is 2-dimensional or not
+    bool m_is2D;
```
```suggestion
    bool m_is_2d;
```
Naming conventions.

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,289 @@
+                                      uint64_t t_ramp)
+        : m_initial_box(box1), m_final_volume(final_volume), m_variant(0, 1, t_start, t_ramp)
+        {
+        m_is2D = m_initial_box->getL().z == 0;
```
```suggestion
        m_is_2d = m_initial_box->getL().z == 0;
```

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,289 @@
+        // current volume = 1 / (current inverse volume)
+        Scalar current_volume = 1 / (s / m_final_volume + (1.0 - s) / m_initial_volume);
+        Scalar L_scale;
+        if (m_is2D)
```
```suggestion
        if (m_is_2d)
```

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,289 @@
+    VariantRamp m_variant; //!< Variant that interpolates between boxes
+    };
+
+namespace detail
+    {
+/// Export Variant classes to Python
+void export_VectorVariantBox(pybind11::module& m);
+void export_VectorVariantBoxConstant(pybind11::module& m);
+
+    } // end namespace detail
```
```suggestion
```
Unused.

joaander commented on hoomd/VectorVariant.cc:
```diff
@@ -0,0 +1,94 @@
+    return (*t)(step);
+    }
+
+void export_VectorVariantBox(pybind11::module& m)
```
```suggestion
void export_VectorVariantBoxClasses(pybind11::module& m)
```
This exports more than `VectorVariantBox`.

joaander commented on hoomd/module.cc:
```diff
@@ -353,6 +354,9 @@ PYBIND11_MODULE(_hoomd, m)
     export_Variant(m);
 
+    // vector variant
+    export_VectorVariantBox(m);
```
```suggestion
    export_VectorVariantBoxClasses(m);
```

joaander commented on hoomd/variant/box.py:
```diff
@@ -0,0 +1,132 @@
+
+    `Constant` returns ``[box.Lx, box.Ly, box.Lz, box.xz, box.xz, box.yz]`` at
+    all time steps.
+
```
```suggestion
```

joaander commented on hoomd/variant/box.py:
```diff
@@ -0,0 +1,132 @@
+
+
+class Interpolate(_hoomd.VectorVariantBoxInterpolate, BoxVariant):
+    """Interpolate between two boxes linearly in time.
```
```suggestion
    """Interpolate between two boxes linearly.
```

joaander commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,289 @@
+    public:
+    /** Construct a VectorVariantBoxInterpolate to interpolate between two boxes linearly in time.
+
+        @param box1 The initial box
+        @param box2 The final box
```
The python code calls these `initial_box` and `final_box`. Please rename all instances of `box1` to `initial_box` and all instances of `box2` to `final_box` for consistency.

joaander commented on hoomd/variant/box.py:
```diff
@@ -0,0 +1,132 @@
+        variant (hoomd.variant.variant_like): A variant used to interpolate
+            between the two boxes.
+
+    `Interpolate` returns the array corresponding to *initial_box* for
+    :math:`t \\leq t_{\\mathrm{start}}` and *final_box* for
+    :math:`t \\geq t_{\\mathrm{start}} + t_{\\mathrm{ramp}}`.
```
This is incorrect, there is no t_start nor t_ramp. Replace this with a copy of the docs from `BoxResize` that explicitly document the interpolation.

joaander commented on hoomd/variant/box.py:
```diff
@@ -0,0 +1,132 @@
+
+
+class InverseVolumeRamp(_hoomd.VectorVariantBoxInverseVolumeRamp, BoxVariant):
+    """Produce box arrays whose inverse volume changes linearly with time.
```
```suggestion
    """Produce box arrays whose inverse volume changes linearly.
```

joaander commented on hoomd/variant/box.py:
```diff
@@ -0,0 +1,132 @@
+
+    ``InverseVolumeRamp`` produces box arrays that correspond to a box whose
+    **inverse volume** (i.e., density for a constant number of particles) varies
+    linearly with time. The shape of the box remains constant, that is, the
```
```suggestion
    linearly. The shape of the box remains constant, that is, the
```

joaander replied:
The ramp is not linear in time. It is constant up to `t_start`, linear between `t_start` and `t_start + t_ramp` and constant after `t_ramp`. Need to explain this concisely in this documentation, possibly with a reference to the scalar `Ramp`.

For example, add a math equation that shows how the volume as a function of time relates to the initial box, final volume, and the ramp function.

joaander reviewed (changes requested):
Thanks! I have some more suggestions to improve this.

Also, I plan to merge this after the 4.5.0 release with the expectation that we can add code using box variants (BoxResize, QuickCompress) in the same release that we introduce the variants.

b-butler commented on hoomd/variant/__init__.py:
```diff
@@ -0,0 +1,21 @@
+provided subclasses.
+"""
+
+from hoomd.variant.scalar import Variant
+from hoomd.variant.scalar import Constant
+from hoomd.variant.scalar import Ramp
+from hoomd.variant.scalar import Cycle
+from hoomd.variant.scalar import Power
```
```suggestion
from hoomd.variant.scalar import (Variant, Constant, Ramp, Cycle, Power)
```

b-butler commented on hoomd/variant/__init__.py:
```diff
@@ -0,0 +1,21 @@
+
+"""Define quantities that vary over the simulation.
+
+A `Variant` object represents a scalar function of the time step. Some
```
```suggestion
A `Variant` object represents a function of the time step. Some
```

b-butler commented on hoomd/VectorVariant.h:
```diff
@@ -0,0 +1,289 @@
+        @param timestep Time step to query.
+        @returns The value of the variant.
+    */
+    virtual array_type operator()(uint64_t timestep)
+        {
+        std::array<Scalar, ndim> ret;
+        ret.fill(0);
+        return ret;
+        }
+    };
```
Shouldn't this be a pure virtual function? Is there any reason to have a default implementation that just returns zeros?

joaander replied:
Exposing pure virtual functions via pybind11 adds a lot of complexity for a minimal benefit in coding style.

b-butler commented on hoomd/VectorVariant.cc:
```diff
@@ -0,0 +1,94 @@
+/// Method to enable unit testing of C++ variant calls from pytest
+std::array<Scalar, 6> testVectorVariantBoxCall(std::shared_ptr<VectorVariantBox> t, uint64_t step)
+    {
+    return (*t)(step);
```
I personally prefer `t->operator()(step)`, but feel free to leave alone.

b-butler commented on hoomd/variant/box.py:
```diff
@@ -0,0 +1,132 @@
+
+    @property
+    def initial_box(self):
+        """hoomd.Box: the initial box."""
```
```suggestion
        """hoomd.Box: The initial box."""
```
Docstrings should start with capital letters. Fix elsewhere too.

b-butler commented on hoomd/variant/box.py:
```diff
@@ -0,0 +1,132 @@
+    pass
+
+
+class Constant(_hoomd.VectorVariantBoxConstant, BoxVariant):
```
The "don't duplicate code" part of me wants to say we should create a decorator which adds the `box` attributes programmatically with documentation to avoid the boilerplate. Don't do that, I just wanted to document it.

b-butler commented on hoomd/pytest/test_box_variant.py:
```diff
@@ -0,0 +1,172 @@
+_test_box1 = hoomd.Box(10, 50, 20, 0.2, 0.4, 0.6)
+_test_box2 = hoomd.Box(16, 25, 36, 0.1, 0.2, 0.3)
+_test_scalar_variant1 = hoomd.variant.Ramp(0, 1, 100, 200)
+_test_scalar_variant2 = hoomd.variant.Ramp(0, 1, 10, 30)
```
Make these fixtures without the underscore.

b-butler commented on hoomd/pytest/test_box_variant.py:
```diff
@@ -0,0 +1,172 @@
+        return hoomd._hoomd._test_vector_variant_box_call(
+            custom_variant, timestep)
+
+    for _t, _f in ((0, 0), (42, 0), (100, 0), (101, 0.01), (150, 0.5),
+                   (175, 0.75), (199, 0.99), (200, 1.0), (250, 1.0), (123456789,
+                                                                      1.0)):
+        test_box.volume = (1 - _f) * _test_box1.volume + _f * final_volume
+        npt.assert_allclose(box_t(custom_variant, _t), box_to_array(test_box))
```
```suggestion
    for t, f in ((0, 0), (42, 0), (100, 0), (101, 0.01), (150, 0.5),
                   (175, 0.75), (199, 0.99), (200, 1.0), (250, 1.0), (123456789,
                                                                      1.0)):
        test_box.volume = (1 - f) * _test_box1.volume + f * final_volume
        npt.assert_allclose(box_t(custom_variant, t), box_to_array(test_box))
```

b-butler commented on hoomd/pytest/test_box_variant.py:
```diff
@@ -0,0 +1,172 @@
+    npt.assert_allclose(get_volume(variant, 110), final_volume)
+    npt.assert_allclose(get_volume(variant, 1010), final_volume)
+    # make sure tilts don't change
+    npt.assert_allclose(box1.tilts, variant(0)[3:])
+    npt.assert_allclose(box1.tilts, variant(5)[3:])
+    npt.assert_allclose(box1.tilts, variant(25)[3:])
+    npt.assert_allclose(box1.tilts, variant(125)[3:])
+    npt.assert_allclose(box1.tilts, variant(625)[3:])
```
```suggestion
    for step in (0, 5, 252 125, 625):
        npt.assert_allclose(box1.tilts, variant(step)[3:])
```
There are a lot of opportunities to shorten the code with for loops.

b-butler reviewed (commented):
We should either rename `variant_like` to `scalar_varient_like` or make very clear it is for scalar variants.

The code looks good generally. Just left a few comments and probably won't be able to get back to this.

tcmoore3 committed:
Merge remote-tracking branch 'origin/trunk-minor' into feature/vector-variant

tcmoore3 committed:
Address reviews

tcmoore3 committed:
formatting

tcmoore3 committed:
flake8

joaander committed:
Apply suggestions from code review

Co-authored-by: Brandon Butler <butlerbr@umich.edu>

joaander committed:
Merge branch 'trunk-minor' into feature/vector-variant

joaander committed:
Apply changes from code review.

joaander committed:
Revise docs.

joaander committed:
Run pre-commit.

joaander committed:
Explicitly document length of return type.

joaander reviewed (approved):
Thanks!

I implemented a few minor suggestions directly on the branch. This is good to merge now.
